### PR TITLE
systemfields: adds a related model field

### LINKED ISSUE #246: systemfields: related object fields issues
Currently, we have the following type of system fields:

ModelField: 
- Purpose: Provides access to a model-level property (e.g. ``record.created`` -> ``record.model.created``) - i.e. created is stored on the same database table as the record json.
- Dumps/loads: ES dumper automatically dumps/loads the values and recreatest the SQLAlchemy model object.

PIDField:
- Purpose: Provides access to a related PersistentIdentifier object (an SQLAlchemy model object stored in a **different** database table), and stores it in the JSON. Ensures efficient handling of related objects
- Dumps/loads: No special handling, because the PersistentIdentifier object is denormalised into the record's JSON.

RelationsField:
- Purpose: Provides access to related Record objects. Stores a list of IDs in the record.
- Dumps/loads: Deserialize certain fields from the Related object into the dump, so it can be recreated.

# Properties/differences

- Type of related objects: A Record (via PIDs/PKs), An SQLAlchemy model (via primary key)
- Place the relation is stored: JSON, model field
- Denormalization: yes/no - Do we denormalize the related object into the record.
- Dumping/loading: Do we dump/load the entire object, only some fields.
- Caching: Do we cache the related object 

### New case

Currently, in Invenio-Vocabularies, a VocabularyRecord related to a VocabularyType via foreign key relationship at the database level. It doesn't currently fit in any of the above categories.

Like PIDField we provide access to a related SQLAlchemy object (not a record), but we don't denormalize the VocabularyType object into the JSON instead the vocabulary type id is stored on the database table.

Like ReleationsField we dump/load the vocabulary type.

So the relationship is somewhat sitting between RelationsFIeld and PIDField.

- VocabularyType
- PIDField

Both the ``RelationsField`` and the ``PIDField`` (invenio-records-resources).

### PR DESCRIPTION
**Pull request by lnielsen**
* Adds a related model field which helps serialize/deserialize a related
  object to/from the record dictionary. The core is extract from the
  PIDField in Invenio-Records-Resources. (closes #246)

### CODE REVIEW

**Comment by lnielsen** [comment]
Related to work being done in https://github.com/inveniosoftware/invenio-vocabularies/pull/17

File: invenio_records/systemfields/relatedmodelfield.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020-2021 CERN.
+# Copyright (C) 2020 Northwestern University.
+#
+# Invenio-Records-Resources is free software; you can redistribute it and/or
```
Comment (ppanero) [inline invenio_records/systemfields/relatedmodelfield.py:6]: minor:

```suggestion
# Invenio-Records is free software; you can redistribute it and/or
```

File: invenio_records/systemfields/relatedmodelfield.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    def pre_commit(self, record):
+        """Called before a record is committed."""
+        # Make sure we serialize/dump the related objet on record.commit() time
+        # as it might have changed.
+        related_obj = getattr(record, self.attr_name)
+        if related_obj is not None:
+            self.set_obj(record, related_obj)
+        else:
+            if self._required:
+                raise RuntimeError("You must provide a related object.")
```
Comment (ppanero) [inline invenio_records/systemfields/relatedmodelfield.py:84]: Simpler? I think is the same?

```suggestion
        elif self._required:
            raise RuntimeError("You must provide a related object.")
```

File: tests/test_systemfields_relatedmodelfield.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    class Record1(Record, SystemFieldsMixin):
+        myobj = RelatedModelField(MyModel, load=my_load, dump=my_dump)
+
+    # Dumping
+    record = Record1({}, myobj=MyModel('a'))
+    assert record['myobj'] == {'custom': 'a'}
+
+    # Loading - we use the loads(dumps()) to load the object
+    record = Record1(dict(record))
+    record.myobj == MyModel('a')
```
Comment (ppanero) [inline tests/test_systemfields_relatedmodelfield.py:106]: missing assert?
```suggestion
    assert record.myobj == MyModel('a')
```

**Review by ppanero — APPROVED** [review]
A couple of minor comments plus:
- There is no tests for `RelatedModelFieldContext` (i.e. `session_merge`).
- The commit history has a release commit in the middle, not sure how you want to handle that since it's an alpha...

Approving because most are minors, just the tests for `RelatedModelFieldContext` if it's not an overkill. If minors get fixed, feel free to merge :)

**Comment by lnielsen** [comment]
Thanks for the review. Fixed comments, and added tests for session merge.

**Pull request merged**